### ISSUE TITLE
xterm sixel support not detected automatically 

### ISSUE THREAD

**Issue by ploum** [hpjansson/chafa#191]
using xterm 384 with xterm -ti vt340

happening:
chafa picture.jpg return pictures in symbols.
chafa -f sixels picture.jpg returns pictures in sixels (perfect quality)

expected:
with "-f sixels", chafa should recognize sixels support and use sixels by default

**Comment by ploum** [comment]
Oops, realized afterward that this was with chafa 1.12. Will test with 1.14 later.

**Comment by hpjansson** [comment]
I think the situation is the same in 1.14. XTerm used to default to having sixel support turned off - you had to enable VT340 support in order to get it, and I couldn't find a way to passively detect when this was the case. It looks like newer XTerm has it turned on. Do you know in which version this was changed, and how old it is?

**Comment by ploum** [comment]
On 24/02/08 11:17, Hans Petter Jansson wrote:
>I think the situation is the same in 1.14. XTerm used to default to having sixel support turned off - you had to enable VT340 support in order to get it, and I couldn't find a way to passively detect when this was the case. It looks like newer XTerm has it turned on. Do you know in which version this was changed, and how old it is?

I had no idea it was turned on by default.

But there’s a very interesting detection mechanism in the lsix project, 
which seems to check for a reply code of "4" to the vt220 escape 
sequence (no, I don’t understand how it works)

```
# IS TERMINAL SIXEL CAPABLE?		# Send Device Attributes
IFS=";?c" read -a REPLY -s -t 1 -d "c" -p $'\e[c' >&2
for code in "${REPLY[@]}"; do
if [[ $code == "4" ]]; then
     hassixel=yup
     break
fi
done

# YAFT is vt102 compatible, cannot respond to vt220 escape sequence.
if [[ "$TERM" == yaft* ]]; then hassixel=yeah; fi
```

Please note that this only works with bash, not in zsh (which may use 
the POSIX `read` version without all those options).

**Comment by hpjansson** [comment]
Looks like I was wrong about XTerm enabling support by default. `xterm -ti vt340` seems to be required still on xterm-389. `lsix` is actively probing the terminal, which we're not yet doing (and which when implemented may still not work in all situations, like when we can't access the controlling terminal).

So the action item here is to implement active probing. I put some of the scaffolding in place in Chafa 1.14, so it may not be that far away.

**Comment by j4james** [comment]
@hpjansson FYI, you can enable the graphics protocols in XTerm without "downgrading" to VT340 by setting the [decGraphicsID](https://invisible-island.net/xterm/manpage/xterm.html#VT100-Widget-Resources:decGraphicsID) resource. That's been an option since patch 357 if the logs are accurate.

**Comment by Safari77** [comment]
Also, when xterm is started from kitty terminal, chafa defaults to using kitty format because $KITTY_PID is set. (Sure, I could make alias to start env -u KITTY_PID xterm ...)

**Comment by hpjansson** [comment]
> Also, when xterm is started from kitty terminal, chafa defaults to using kitty format because $KITTY_PID is set. (Sure, I could make alias to start env -u KITTY_PID xterm ...)

Yes, this is a weakness with env var-based detection. I don't think there's anything we can reasonably do about it. Open to ideas, though. One fairly brutal hack I've thought of is to look at the process tree to see what we're running under, but that doesn't work in the remote case, obviously.
